keps/sig-storage/20190408-volume-scheduling-limits.md

```diff
@@ -0,0 +1,294 @@
+  * To get rid of prefix + SHA for `ResourceName` of CSI volumes.
+  * So in-tree volume plugin can share limits with CSI driver that uses the same storage backend.
+
+* Kubelet will create `CSINode` instance during initial node registration together with `Node` object.
+  * Limits of each in-tree volume plugin will be added to `CSINode.status.allocatable`.
+    * Limit for in-tree volumes will be added by kubelet during CSINode creation. Name of corresponding CSI driver will be used as key in `CSINode.status.allocatable` and it will be discovered using [CSI translation library](https://github.com/kubernetes/kubernetes/tree/master/staging/src/k8s.io/csi-translation-lib). If the library does not support migration of an in-tree volume plugin, the volume plugin has no limit.
+    * If a CSI driver is registered for an in-tree volume plugin and it reports a different volume limit than in-tree volume plugin, the limit reported by CSI driver is used and kubelet logs a warning.
+  * User may NOT change `CSINode.status.allocatable` to override volume plugin / CSI driver values, e.g. to "reserve" some attachment to the operating system. Kubelet will periodically reconcile `CSINode` and overwrite the value.
+    * Especially, `kubelet --kube-reserved` or `--system-reserved` cannot be used to "reserve" volumes for kubelet or the OS. It is not possible with current kubelet and this KEP does not change it. We expect that CSI drivers will have configuration options / cmdline arguments to reserve some volumes and they will report their limit already reduced by that reserved amount.
+
+  * Kubelet will continue filling `Node.status.allocatable` and `Node.status.capacity` for both in-tree and CSI volumes during deprecation period. After the deprecation period, it will stop using them completely.
+    * Scheduler (all its storage predicates) will ignore `Node.status.allocatable` and `Node.status.capacity` if `CSINode.status.allocatable` is present.
+    * If `CSINode.status.allocatable` (or whole `CSINode`) is missing, scheduler falls back to `Node.status.allocatable`. This solves version skew between old kubelet (using `Node.status`) and new scheduler.
+    * After deprecation period, scheduler won't schedule any pods that use volumes to a node with missing `CSINode` instance. It is expected that it happens only during node registration when `Node` exists and `CSINode` doesn't and it self-heals quickly.
+
+* `CSINode.status.allocatable` is array of limits. Following combinations are possible:
+
+|  `Volumes` | Description  |
+|  --------- | ------------ |
+|  0         | plugin / CSI driver exists and has zero limit, i.e. can attach no volumes |
+|  X>0       | plugin / CSI driver exists and can attach X volumes (where X > 0) |
+|  X<0       | negative values are blocked by validation
+| Driver is missing in `CSINode.status.allocatable`   | there is no limit of volumes on the node* |
```

Comment:
When does not happen and what does it mean that "driver" does not exist in `CSINode.status.allocatable`?

Reply:
> what does it mean that "driver" does not exist in CSINode.status.allocatable

`CSINode.status.allocatable` is a `map[DriverName]VolumeLimits`. If a driver is missing there, it can mean two things:
  * Driver is not installed on the node (or it is installed and informers are slow to propagate it to all components)
  * Driver is installed on the node and has no limits.
In both cases, scheduler expects that there is no limit.

Differentiation between these two cases is discussed here: https://github.com/kubernetes/enhancements/pull/942#discussion_r276620931